HW2 m2-prepare

kondury opened the pull request:
added README.md
added project description
added devops infrastructure description and schemas
added architecture description and schemas
added user stories
added entities and functions description
added frontend layout
added deployment infrastructure
added acceptance test with wiremock

kondury committed:
HW2 m2-prepare
added README.md
added project description
added devops infrastructure description and schemas
added architecture description and schemas
added entities and functions description
added frontend layout
added deployment infrastructure
added acceptance test with wiremock

svok commented on docs/01-marketing/user-stories.md:
```diff
@@ -0,0 +1,126 @@
+## 1. Запуск процесса изучения невыученных карточек
+
+**В качестве** пользователя сервиса \
+**Я хочу** изучать новые и повторять невыученные карточки \
```
Карточки - это одно из тысяч возможных решений.
Надо понять какая именно цель у пользователя.
Например, хочу изучить английский язык.

kondury replied:
в описании профилей я описал, что в рамках MVP рассматриваем пользователя, который готовится к техническому собеседованию.

svok commented on docs/01-marketing/user-stories.md:
```diff
@@ -0,0 +1,126 @@
+### 1.2. Сценарий: повторение ранее пройденных карточек. 
+**Дано:** я нахожусь на главном экране \
+**Когда** я хочу повторять ранее пройденные карточки, я нажимаю кнопку "Повторение", \
+**тогда** запускается повторение ранее пройденных карточек \
```
Типичная ошибка, которой страдает большинство аналитиков - это писать аналитику под конкретное решение, а не проблему. Надо сосредоточиться именно на потребностях пользователя, а не на том, что мы ему предлагаем с карточками возиться.
Если распишите с т.з. проблематики, возможно выплывает другое решение, а не карточки.

kondury replied:
в рамках MVP рассматриваю конкретный инструмент "карточки" и систему повторений. В сценариях я конкретизировал запросы под предполагаемый фронтенд. В историях конкретики меньше, но она всё же завязана на карточки, так как это основная идея приложения.

svok commented on docs/04-architecture/03-api.md:
```diff
@@ -0,0 +1,60 @@
+   6. Среднее количество просмотров в день
+   7. Среднее количество просмотров до перевода карточки в состояние выучено
+   8. Среднее количество выученных карточек в день
+3. CRUDS для карточек (опционально)
```
разве вы сможете сделать повторение без карточки?

kondury replied:
Пока это только сущности домена.
Так как схему БД и внутреннюю модель пока не рассматриваем, вопрос открытый и будет уточняться на следующих этапах. Для MVP предполагаю жесткую загрузку карточек в систему (один общий список). Поэтому работа с карточками может свестись к выборке каких-то проекций DTO из "коробок"+"карточки". Если "коробка" переместится в "карточку", как атрибут, то будет CRUD карточки, но не будет "коробки".

svok commented on docs/04-architecture/03-api.md:
```diff
@@ -0,0 +1,60 @@
+   2. Регистрация повторений
+      1. Перемещение в коробку "выученных"
+      2. Перемещение в коробку "на повторении"
+2. Анализ обучения, статистические методы:
```
Это функции такие?  У них конкретный эндпоинт предполагается?
`/cards/unstudiedCount`
Так?
Думаю, это не функции, а аналитические данные, которые мы будем рассматривать в разделе мониторинга. А здесь они не нужны

kondury replied:
ясно. Так как статистика нужна, пока оставлю здесь. Потом перемещу описание в соответствующий раздел.

svok commented on docs/04-architecture/03-api.md:
```diff
@@ -0,0 +1,60 @@
+Атрибуты:
+1. id
+2. front (лицевая сторона)
+3. back (обратная сторона)
```
О чем речь? Это текст карточки? Изображение? Документ?

svok replied:
Карточка принадлежит системе или у нее есть конкретный владелец?

kondury replied:
представление данных - текст. Делал по аналогии с учебным проектом, в котором ни слова о типах данных нет.

kondury replied:
в MVP будет один набор карточек для всех пользователей

svok commented on docs/04-architecture/03-api.md:
```diff
@@ -0,0 +1,60 @@
+Атрибуты:
+1. id
+2. name (имя): имя коробки
+3. cards (карточки): карточки, лежащие в коробке
```
Что это за карточки в коробке? Это список айдишников? А если в коробке будет 20 тыс карточек? Мы прямо все 20 тыс полжим в одно поле?

Обычно либо делают таблицу-связку, в которой указывают какая карточка в каких коробках лежит, либо прямо в самой карточке указывают идентификатор коробки. Можно и список идентификаторов коробок в карточке указать. Коробок же значительно меньше, чем карточек.

kondury replied:
Описывал сущности, а не их реализацию. В ORM, вероятно, будет двухсторонняя связь, с ленивой инициализацией или что-то в этом духе. Да, коробка вполне вероятно выродится в value class и будет использоваться как атрибут карточки. Потенциально, коробок может быть больше, поэтому решение отложено до начала реализации.

svok commented on docs/04-architecture/03-api.md:
```diff
@@ -0,0 +1,60 @@
+
+Атрибуты:
+1. id
+2. card (карточка)
```
Вся карточка или идентификатор?

kondury replied:
На уровне БД, скорее всего id.  На уровне приложения будет зависеть от фреймворка и других соображений.

svok commented on docs/04-architecture/03-api.md:
```diff
@@ -0,0 +1,60 @@
+4. box after (результат: коробка, в которую помещается карточка после повторения)
+5. started time (время начала повторения)
+6. finished time (время завершения, регистрация результата самопроверки)
+
```
Вы здесь много сущностей описали.  Но так и не ясно что будет в вашем MVP.
Все сущности вряд ли вы успеете сделать, если только вы не свободны от основной работы.

svok reviewed (commented):
Константин, отличная работа.
Рекомендую сократить объем работ до одной-двух сущностей и ускориться.
Если отстанете, наверстать будет тяжело.
А среди тех, кто переводится на следующий поток, очень низкий процент защит.

kondury committed:
HW2 m2-prepare - fixes after review
minor fixes in README.md
03-api.md:
- excluded progress statistics
- made API description more exact
- refactored entities into two most significant
